src/include/txn_inline.h

```diff
@@ -716,8 +716,9 @@ __txn_visible_id(WT_SESSION_IMPL *session, uint64_t id)
     if (txn->isolation == WT_ISO_READ_UNCOMMITTED)
         return (true);
 
-    /* Otherwise, we should be called with a snapshot. */
-    WT_ASSERT(session, F_ISSET(txn, WT_TXN_HAS_SNAPSHOT) || session->dhandle->checkpoint != NULL);
+    /* Otherwise, we need a snapshot; if we don't have one, assume a test for global visibility. */
+    if (!F_ISSET(txn, WT_TXN_HAS_SNAPSHOT))
+        return (__txn_visible_all_id(session, id));
```

Comment:
I don't have context on the actual failure scenario but shouldn't the caller know if it doesn't have a snapshot and call for a visible_all check intentionally?

Reply:
Yes and no: the top-level caller certainly knows, but when you're walking the tree, the mid-level tree walk functions call:
```
__wt_delete_page_skip() ->
    __wt_page_del_active() ->
        __wt_txn_visible()
```
and the tree walk functions don't know if its caller has a snapshot or not. We could pass that information in (but the tree walk function API is pretty complicated and flag-heavy already).

Here's the link to that call: https://github.com/wiredtiger/wiredtiger/blob/develop/src/btree/bt_walk.c#L426-L431

There's some discussion in WT-8604 if you want more details, @hbokhari raised exactly that point.